cw/cw305/ceca.py

```diff
@@ -0,0 +1,574 @@
+        max_std: Allowed number of standard deviations from the mean trace for
+            filtering noisy traces.
+        num_workers: Number of workers to use for processing traces.
+
+    Returns:
+        Recovered key if the attack was successful, ``None`` otherwise.
+    """
+    project = cw.open_project(project_file)
+
+    # Check arguments
+    num_total_traces = len(project.waves)
+    if num_traces > num_total_traces:
+        raise ValueError(f'Invalid num_traces: {num_traces} (must be greater '\
+                         f'than {num_total_traces})')
+    last_sample = len(project.waves[0]) - 1
+    if min(attack_window) < 0 or max(attack_window) > last_sample:
+        raise ValueError(f'Invalid attack window: {attack_window} (must be '\
+                         f'in [0, {last_sample}])')
+    if max_std <= 0:
+        raise ValueError(f'Invalid max_std: {max_std} (must be greater than '\
+                         'zero)')
+    if num_workers <= 0:
+        raise ValueError(f'Invalid num_workers: {num_workers} (must be greater'\
+                         ' than zero)')
+
+    # Instantiate workers
+    def worker_trace_slices():
+        """Determines the traces of each worker.
+
+        Assigns the remainder, if any, to the first worker.
+        """
+        traces_per_worker = int(num_traces / num_workers)
+        first_worker_num_traces = traces_per_worker + num_traces % num_workers
+        yield slice(0, first_worker_num_traces)
+        for trace_begin in range(first_worker_num_traces, num_traces,
+                                 traces_per_worker):
+            yield slice(trace_begin, trace_begin + traces_per_worker)
+
+    attack_window = slice(*attack_window)
+    workers = [
+        TraceWorker.remote(project_file, trace_slice, attack_window,
+                           attack_direction)
+        for trace_slice in worker_trace_slices()
+    ]
+    assert len(workers) == num_workers
+    # Compute mean and standard deviation.
+    mean, std_dev = compute_mean_and_std(workers)
+    # Filter noisy traces.
+    orig_num_traces = num_traces
+    num_traces = filter_noisy_traces(workers, mean, std_dev, max_std)
+    logging.info(f'Will use {num_traces} traces '\
+                 f'({100*num_traces/orig_num_traces:.1f}% of all traces)')
+    # Mean traces for all values of all plaintext bytes.
+    mean_plaintext_traces = compute_mean_plaintext_traces(workers)
+    # Guess the differences between key bytes.
+    pairwise_diffs_scores = compute_pairwise_diffs_and_scores(
+        mean_plaintext_traces)
+    diffs = find_best_diffs(pairwise_diffs_scores)
+    logging.info(f'Difference values (delta_0_i): {diffs}')
+    # Recover the key.
+    key = recover_key(diffs, attack_direction, project.textins[0],
+                      project.textouts[0])
+    if key is not None:
+        logging.info(f'Recovered AES key: {bytes(key).hex()}')
+    else:
+        logging.error(f'Failed to recover the AES key')
```

Comment:
I think so, but based on my limited experiments this sometimes can be a red herring because of the noise and the way we utilize difference values. That's why I didn't include it in this PR (also to keep things simple) but I don't think it will introduce a noticeable overhead and should be easy to add. Created #30.